File: node/ts/zkgroup/calllinks/CallLinkAuthCredential.ts
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

import { randomBytes } from 'node:crypto';

import ByteArray from '../internal/ByteArray.js';
import { RANDOM_LENGTH } from '../internal/Constants.js';
import * as Native from '../../Native.js';

import CallLinkSecretParams from './CallLinkSecretParams.js';
import CallLinkAuthCredentialPresentation from './CallLinkAuthCredentialPresentation.js';
import GenericServerPublicParams from '../GenericServerPublicParams.js';
import { Aci } from '../../Address.js';

export default class CallLinkAuthCredential extends ByteArray {
  private readonly __type?: never;

  constructor(contents: Uint8Array) {
    super(contents, Native.CallLinkAuthCredential_CheckValidContents);
  }

  present(
    userId: Aci,
    redemptionTime: number,
    serverParams: GenericServerPublicParams,
    callLinkParams: CallLinkSecretParams
  ): CallLinkAuthCredentialPresentation {
    const random = randomBytes(RANDOM_LENGTH);
    return this.presentWithRandom(
      userId,
      redemptionTime,
      serverParams,
      callLinkParams,
      random
    );
  }

  presentWithRandom(
    userId: Aci,
    redemptionTime: number,
    serverParams: GenericServerPublicParams,
    callLinkParams: CallLinkSecretParams,
    random: Uint8Array
  ): CallLinkAuthCredentialPresentation {
    return new CallLinkAuthCredentialPresentation(
      Native.CallLinkAuthCredential_PresentDeterministic(
        this.contents,
        userId.getServiceIdFixedWidthBinary(),
        redemptionTime,
        serverParams.contents,
        callLinkParams.contents,
        random
      )
    );
  }
}

File: java/shared/java/org/signal/libsignal/zkgroup/calllinks/CallLinkAuthCredential.java
//
// Copyright 2023 Signal Messenger, LLC.
// SPDX-License-Identifier: AGPL-3.0-only
//

package org.signal.libsignal.zkgroup.calllinks;

import org.signal.libsignal.protocol.ServiceId.Aci;
import org.signal.libsignal.zkgroup.GenericServerPublicParams;
import org.signal.libsignal.zkgroup.InvalidInputException;
import org.signal.libsignal.zkgroup.internal.ByteArray;
import org.signal.libsignal.internal.Native;

import java.security.SecureRandom;
import java.time.Instant;

import static org.signal.libsignal.zkgroup.internal.Constants.RANDOM_LENGTH;

public final class CallLinkAuthCredential extends ByteArray {

  public CallLinkAuthCredential(byte[] contents) throws InvalidInputException {
    super(contents);
    Native.CallLinkAuthCredential_CheckValidContents(contents);
  }

  public CallLinkAuthCredentialPresentation present(Aci userId, Instant redemptionTime, GenericServerPublicParams serverParams, CallLinkSecretParams callLinkParams) {
    return present(userId, redemptionTime, serverParams, callLinkParams, new SecureRandom());
  }

  public CallLinkAuthCredentialPresentation present(Aci userId, Instant redemptionTime, GenericServerPublicParams serverParams, CallLinkSecretParams callLinkParams, SecureRandom secureRandom) {
    byte[] random      = new byte[RANDOM_LENGTH];
    secureRandom.nextBytes(random);

    byte[] newContents = Native.CallLinkAuthCredential_PresentDeterministic(getInternalContentsForJNI(), userId.toServiceIdFixedWidthBinary(), redemptionTime.getEpochSecond(), serverParams.getInternalContentsForJNI(), callLinkParams.getInternalContentsForJNI(), random);

    try {
      return new CallLinkAuthCredentialPresentation(newContents);
    } catch (InvalidInputException e) {
      throw new AssertionError(e);
    }
  }

}
